JAMES-2420 DLP rule store

chibenwa opened the pull request.

trantienduchn commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRule.java:
```diff
@@ -0,0 +1,184 @@
+        private final boolean contentTargeted;
+
+        public Targets(boolean senderTargeted, boolean recipientTargeted, boolean contentTargeted) {
+            this.senderTargeted = senderTargeted;
```
do we need to check if 3 targets are all `false` ?

chibenwa replied:
Why not. It would test explicitly default values and it's always better.

rouazana replied:
that could be a not so much bad way to deactivate (temporarily) a rule.

trantienduchn commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRule.java:
```diff
@@ -0,0 +1,184 @@
+public class DLPRule {
+
+    public static class Builder {
+        public static final boolean NOT_TARGETTED = false;
```
`NOT_TARGETED` I think 😸

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRule.java:
```diff
@@ -0,0 +1,184 @@
+        private Optional<Boolean> targetsSender;
+        private Optional<Boolean> targetsRecipients;
+        private Optional<Boolean> targetsContent;
+        private Optional<String> explanation;
```
description ?

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRulesStore.java:
```diff
@@ -0,0 +1,41 @@
+
+public interface DLPRulesStore {
+
+    Map<DLPRuleId, DLPRule> retrieveRules(Domain domain);
```
we don't care  that much about Id, it should be a field of DLPRule and you should have a ```List<DLPRule> list()``` method IMO

rouazana replied:
I agree

chibenwa replied:
That means that the user have to set the id as part of the request?

That seems like a bad idea to me! I'd rather not let him generate the id...

rouazana replied:
I think we can remove the id field for now, or let the user generate it.

mbaechler replied:
> That seems like a bad idea to me! I'd rather not let him generate the id...

Why ? It's often the best solution to ensure idempotent property.

rouazana commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRulesStore.java:
```diff
@@ -0,0 +1,41 @@
+
+    Optional<DLPRule> retrieveRule(Domain domain, DLPRuleId dlpRuleId);
+
+    DLPRuleId store(Domain domain, DLPRule rule);
```
we cannot allow to store only one Rule, all rules must be stored at once, else you need to handle ordering

chibenwa replied:
Why would you need ordering? Isn't it a AND between rules & AND commutes?

rouazana replied:
I'm not sure it's absolutely needed. But once you get a list of rules, you can be tempted to order them. For example for performance reasons.

chibenwa commented:
I push here the event sourcing version of this.

We simplified the API to not handle single record CRUD, adapted the tests, provide an event sourcing version, and dropped the memory one.

trantienduchn commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregateId.java:
```diff
@@ -0,0 +1,61 @@
+    private final Domain domain;
+
+    public DLPRuleAggregateId(Domain domain) {
+        this.domain = domain;
```
It needs `Preconditions` check

chibenwa commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/EventSourcingDLPRuleStore.java:
```diff
@@ -0,0 +1,104 @@
+        eventSourcingSystem.dispatch(new ClearCommand(domain));
+    }
+
+    private Optional<Event> getLastEvent(Domain domain) {
```
I find it useful for reading to have the History type thus I'd like to keep it as is.

chibenwa commented:
:green_apple:  (yoda failure on last commit, last commit is just about visibility thus do not impact deployment tests, hence I concider it :green_apple:  )

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/SimpleDLPRuleId.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public String serialize() {
+        return null;
```
why ?

chibenwa replied:
This class is no longer used, CF @rouazana  comments.

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregateId.java:
```diff
@@ -0,0 +1,61 @@
+
+public class DLPRuleAggregateId implements AggregateId {
+    private static final String SEPARATOR = "/";
+    private static final String PREFIX = "QuotaThreasholdEvents";
```
probably not the right prefix

chibenwa replied:
Already fixed

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregateId.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public String asAggregateKey() {
+        return Stream.of(
```
what about ```return PREFIX + SEPARATOR + domain.asString()``` ?

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/commands/ClearCommandHandler.java:
```diff
@@ -0,0 +1,54 @@
+    public List<? extends Event> handle(ClearCommand clearCommand) {
+        DLPRuleAggregateId aggregateId = new DLPRuleAggregateId(clearCommand.getSenderDomain());
+        History eventsOfAggregate = eventStore.getEventsOfAggregate(aggregateId);
+        return ImmutableList.of(new ClearEvent(
```
At this point, you have to load the aggregate and ask it to take the decision

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/EventSourcingDLPRuleStore.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public Map<DLPRuleId, DLPRule> retrieveRules(Domain domain) {
+        return getLastEvent(domain)
```
It's not the way it should be done.
Either you load the aggregate and ask it the list or you build a projection with a Listener and you get the list from this projection.

chibenwa replied:
Let's keep it simple and ask the aggregate the list (the aggregate will be reading the last event)

mbaechler commented on server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/commands/StoreCommandHandler.java:
```diff
@@ -0,0 +1,55 @@
+    public List<? extends Event> handle(StoreCommand storeCommand) {
+        DLPRuleAggregateId aggregateId = new DLPRuleAggregateId(storeCommand.getDomain());
+        History eventsOfAggregate = eventStore.getEventsOfAggregate(aggregateId);
+        return ImmutableList.of(new StoreEvent(
```
you have to rely on the aggregate to take this decision

rouazana commented on server/data/data-api/src/main/java/org/apache/james/dlp/api/DLPRule.java:
```diff
@@ -0,0 +1,185 @@
+        }
+
+        public DLPRule build() {
+            Preconditions.checkState(expression.isPresent(), "`expression` in mandatory");
```
s/in/is/

rouazana commented on server/data/data-api/src/test/java/org/apache/james/dlp/api/DLPRuleStoreContract.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Test
+    default void clearShouldNotFailWhenDomainDoNotExist(DLPRulesStore dlpRulesStore) {
```
s/Do/Does

rouazana commented on server/data/data-api/src/test/java/org/apache/james/dlp/api/DLPRuleTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Test
+    void expressionShouldIsTheOnlyMandatoryField() {
```
s/Is/Be/

rouazana commented on server/data/data-api/src/test/java/org/apache/james/dlp/api/DLPRuleTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Test
+    void targetsContentShouldBeReportedInTargets() {
```
maybe you can add a test with the three targets defined at the same time

rouazana commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/EventSourcingDLPRuleStore.java:
```diff
@@ -0,0 +1,81 @@
+
+import com.google.common.collect.ImmutableSet;
+
+public class EventSourcingDLPRuleStore implements DLPRulesStore {
```
s/EventSourcingDLPRuleStore/EventSourcingDLPRulesStore ? (missing s)

rouazana commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregateId.java:
```diff
@@ -0,0 +1,60 @@
+
+import com.google.common.base.Preconditions;
+
+public class DLPRuleAggregateId implements AggregateId {
```
also DLPRule**s** ?

rouazana commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregateId.java:
```diff
@@ -0,0 +1,60 @@
+
+public class DLPRuleAggregateId implements AggregateId {
+    private static final String SEPARATOR = "/";
+    private static final String PREFIX = "DLPRule";
```
idem?

mbaechler commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/EventSourcingDLPRuleStore.java:
```diff
@@ -63,12 +59,13 @@ public EventSourcingDLPRuleStore(EventStore eventStore) {
+
+        DLPRuleAggregateId aggregateId = new DLPRuleAggregateId(domain);
+
+        return DLPRuleAggregate.load(
```
What about DLPDomainRules ?

mbaechler commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregate.java:
```diff
@@ -0,0 +1,72 @@
+        return ImmutableList.of(new StoreEvent(aggregateId, history.getNextEventId(), rules));
+    }
+
+    private Optional<Event> getLastEvent() {
```
This looks like a good candidate to move into history

mbaechler commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregate.java:
```diff
@@ -0,0 +1,72 @@
+    public Stream<DLPRule> retrieveRules() {
+        return getLastEvent()
+            .filter(event -> event instanceof StoreEvent)
+            .map(event -> (StoreEvent) event)
```
StoreEvent.class::cast ?

mbaechler commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregate.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public List<Event> clear() {
+        return ImmutableList.of(new ClearEvent(aggregateId, history.getNextEventId()));
```
Event are always from the past, it should be named DLPRuleRemoved or something like that.
Moreover, it should not be confused with the `Command` you received, events generated by aggregates are result of decision. I propose ```DLPRuleRemoved(aggregateId, history.getNextEventId(), existingRules)``` or ```existingRules.stream.map(event -> new DLPRuleRemoved(aggregateId, history.getNextEventId(), event))```

mbaechler commented on server/data/data-library/src/main/java/org/apache/james/dlp/eventsourcing/aggregates/DLPRuleAggregate.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public List<Event> store(List<DLPRule> rules) {
+        return ImmutableList.of(new StoreEvent(aggregateId, history.getNextEventId(), rules));
```
same comment then behind : ```DLPRuleAdded(aggregateId, history.getNextEventId(), rule)```

mbaechler commented:
see https://github.com/linagora/james-project/pull/1478
